### PR TITLE
Unit tests and CI on MSVC/Windows

### PR DESCRIPTION
**Pull request by PascalGuenther**
- Unit tests use std::string instead of std::vector<char>
- Unit test can be run in Microsoft Visual Studio 2019

### CODE REVIEW

**Review by platisd — CHANGES_REQUESTED** [review]
Thanks for your contributions, they look on point!

Some remarks:
* Regarding the strings:
  Please don't replace the vector of characters since it's part of "[Homework](https://github.com/platisd/reusable-testable-arduino-tutorial#homework)":

> Think of a better way to return an Arduino String in a platform-agnostic way.

* Regarding MSVC compatibility:
  Looks good, I have little experience with Windows but the changes look reasonable. I wonder, is it easy to add a CI build for MSVC so to verify the MSVC-specific functionality? If this was to be merged, someone would have to try things out manually to verify they work.

**Comment by PascalGuenther** [comment]
Thanks for your comments. I have set up a CI build for Windows. Regarding std::string, I will open a new PR targeting the solutuions branch.

File: .github/workflows/firmware-ci.yml [hunk 1]
```diff
@@ -28,30 +28,48 @@ jobs:
...
-            cxx: clang++-9
+        os: [ubuntu-18.04, ubuntu-18.04, windows-latest]
+        include:
+        - os: ubuntu-18.04
+          cc: gcc-9
+          cxx: g++9
+
+        - os: ubuntu-18.04
+          cc: gcc-9
+          cxx: g++-9
```
Comment (platisd) [inline .github/workflows/firmware-ci.yml:57]: I guess you meant `clang-9`? :)

File: .github/workflows/firmware-ci.yml [hunk 2]
```diff
@@ -28,30 +28,48 @@ jobs:
...
+        run: mkdir -p ${{ env.build_dir }}
+      - name: Create unit test build directory
+        if: runner.os == 'Windows'
+        run: mkdir -Force ${{ env.build_dir }}
       - name: Configure unit tests
-        run: cd ${build_dir} && cmake ..
+        run: cd ${{ env.build_dir }} && cmake ..
+      - name: Build unit tests
+        if: runner.os == 'Linux'
+        run: cmake --build ${{ env.build_dir }} -- -j$(nproc)
```
Comment (platisd) [inline .github/workflows/firmware-ci.yml]: I guess `nproc` doesn't work on Windows.Then we can remove it to keep this CI configuration the same across OS's.
In other words, run the same command on both Ubuntu and Windows and skip `nproc`

File: .github/workflows/firmware-ci.yml [hunk 3]
```diff
@@ -28,30 +28,48 @@ jobs:
...
+          cc: gcc-9
+          cxx: g++9
+
+        - os: ubuntu-18.04
+          cc: gcc-9
+          cxx: g++-9
+
+        - os: windows-latest
+          cc: cl
+          cxx: cl
```
Comment (platisd) [inline .github/workflows/firmware-ci.yml:65]: Cool! :+1:

**Review by platisd — CHANGES_REQUESTED** [review]

File: .github/workflows/firmware-ci.yml (hunk 1 above)
Comment (PascalGuenther) [reply .github/workflows/firmware-ci.yml:57]: yes :)

File: .github/workflows/firmware-ci.yml (hunk 2 above)
Comment (PascalGuenther) [reply .github/workflows/firmware-ci.yml]: indeed, Windows doens't seem to supprt this

**Comment by PascalGuenther** [comment]
PR updated according to your comments

**Comment by platisd** [comment]
Looks really good! I admire your passion for detail! :} :+1:

**Pull request merged**